elks/arch/i86/mm/segment.S

```diff
@@ -1,11 +1,88 @@
 /*
- *	Assembly user access routines for the kernel.
+ *  Assembly memory access routines for the kernel.
  */
  
 #include <arch/asm-offsets.h>
 
 	.text
 
+/*
+ *  void fmemset (word_t off, seg_t seg, byte_t val, word_t count);
+ *  segment after offset to allow LES from the stack
+ *  byte_t pushed as word_t by BCC
+ */
+
+	.globl	_fmemset
+
+_fmemset:
+	push   bp
+	mov    bp,sp
+	push   es
+	push   di
+	push   cx
+	les    di,[bp+4]           ; arg0+1: far pointer
+	mov    ax,[bp+8]           ; arg2: value
+	mov    cx,[bx+10]          ; arg3: byte count
+	cld
+	test   cx,#1               ; optimize for word multiple
+	jz     fmemset_word
+	rep
+	stosb
+	jmp    fmemset_exit
+
+fmemset_word:
+	shr    cx,#1
+	mov    ah,al
+	rep
+	stosw
+
+fmemset_exit:
+	pop    cx
+	pop    di
+	pop    es
+	pop    bp
+	ret
+
+/*
+ * void fmemcpy (word_t dst_off, seg_t dst_seg, word_t src_off, seg_t src_seg, word_t count);
+ * segment after offset to allow LDS & LES from the stack
+ */
+
+	.globl	_fmemcpy
+
+_fmemcpy:
+	push   bp
+	mov    bp,sp
+	push   es
+	push   ds
+	push   di
+	push   si
+	push   cx
+	les    di,[bp+4]           ; arg0+1: far destination pointer
```

Comment:
Will this conflict with ia16-gcc?

Reply:
This is a portable implementation that works on all the 16-bits compilers (no more assumption on the register usage, except AX for the return value).